### PR TITLE
fix(tests): pin MCP live-gateway client URLs to IPv4 to avoid localhost IPv6 stall

### LINKED ISSUE #5162: [BUG]: MCP RBAC + protocol e2e gates fail on the multi-replica stack
### Summary

The MCP protocol gates (`make test-mcp-rbac` and `make test-mcp-protocol-e2e`) fail on the
documented 3-replica `make testing-up` stack. The failure is deterministic (`test-mcp-rbac` ran
twice with identical results) and reproduces identically on v1.0.2 and v1.0.3.

### Steps to reproduce

1. Build the gateway image:

       make docker

2. Bring up the documented 3-replica testing stack:

       make testing-down compose-clean testing-up

3. Wait for the stack to be healthy (3 gateway replicas; `http://localhost:8080/health` returns 200).

4. Export the required env (`make` runs pytest via `uv run`, which does not load `.env`):

       export JWT_SECRET_KEY=$(grep -E '^JWT_SECRET_KEY=' .env | cut -d= -f2-)
       export PLATFORM_ADMIN_EMAIL=$(grep -E '^PLATFORM_ADMIN_EMAIL=' .env | cut -d= -f2-)
       export MCP_CLI_BASE_URL=http://localhost:8080

5. Run the gates:

       make test-mcp-rbac test-mcp-protocol-e2e

### Results (summary)

| Target | Runs | Passed | Failed / Errors | Skipped | Rerun |
|---|---|---|---|---|---|
| `test-mcp-rbac` | 2 (identical) | 13 | 27 failed | 0 | 23 |
| `test-mcp-protocol-e2e` | 1 | 2 | 19 errors | 1 | 7 |

Same result on v1.0.2 (`test-mcp-rbac`: 26 failed / 14 passed; `test-mcp-protocol-e2e`: 19 errors / 2 passed / 1 skipped), so this is pre-existing, not introduced in v1.0.3.

<details>
<summary>Full §5.4 failure report (counts + failing test list, Release/v1.0.3)</summary>

Stack: `make testing-up` (default `GATEWAY_REPLICAS=3`); all gateway replicas + nginx healthy; `/health` = 200; required env exported (`JWT_SECRET_KEY`, `PLATFORM_ADMIN_EMAIL`, `MCP_CLI_BASE_URL`).

| Target | Runs | Passed | Failed / Errors | Skipped | Rerun | Duration |
|---|---|---|---|---|---|---|
| `test-mcp-rbac` | 2 (identical) | 13 | 27 failed | 0 | 23 | ~340s |
| `test-mcp-protocol-e2e` | 1 | 2 | 19 errors | 1 | 7 | ~146s |

`test-mcp-rbac` was run twice and produced identical results both times (same counts, same failing tests), so the failure is deterministic, not flaky.

The dominant failure on both targets (every test that opens an MCP client session) occurs at the streamable HTTP `/mcp/` initialize handshake (`fastmcp/client/client.py:532`).

#### `test-mcp-rbac` — 27 failing (reproduced across both runs)

One distinct RBAC-visibility assertion; the other 26 are the MCP session-init failure.

Visibility assertion (separate failure, not session-init):
- `TestServerVisibilityViaAPI::test_admin_sees_public_and_team_via_http` — `AssertionError: PR #4341: admin via HTTP must NOT see private server (own-private collapsed by get_scoped_resource_access_context)`

MCP session init failures (26):
- `TestMcpToolsVisibilityByRole::test_admin_sees_all_tools`
- `TestMcpToolsVisibilityByRole::test_developer_sees_public_and_team_tools`
- `TestMcpToolsVisibilityByRole::test_viewer_sees_public_and_team_tools`
- `TestMcpToolsVisibilityByRole::test_outsider_sees_only_public_tools`
- `TestMcpToolsVisibilityByRole::test_team_admin_sees_public_and_team_tools`
- `TestMcpResourcesPromptsByRole::test_admin_resources`
- `TestMcpResourcesPromptsByRole::test_admin_prompts`
- `TestMcpResourcesPromptsByRole::test_developer_resources_and_prompts`
- `TestMcpResourcesPromptsByRole::test_outsider_resources_and_prompts`
- `TestMcpToolCallByRole::test_admin_calls_tool_success`
- `TestMcpToolCallByRole::test_developer_can_execute_on_default_endpoint`
- `TestMcpToolCallByRole::test_team_admin_can_execute_on_default_endpoint`
- `TestMcpToolCallByRole::test_viewer_can_execute_on_default_endpoint`
- `TestMcpScopedTokenPermissions::test_tools_read_only_token_can_initialize`
- `TestMcpScopedTokenPermissions::test_read_execute_token_can_initialize`
- `TestMcpScopedTokenPermissions::test_unscoped_admin_token_can_call_tools`
- `TestMcpSSETransport::test_sse_tools_discoverable`
- `TestMcpSSETransport::test_sse_get_system_time`
- `TestMcpSSETransport::test_sse_convert_time`
- `TestMcpSSETransport::test_sse_resources_discoverable`
- `TestMcpSSETransport::test_sse_prompts_discoverable`
- `TestMcpPerServerEndpoint::test_public_token_accesses_public_server`
- `TestMcpPerServerEndpoint::test_team_member_accesses_team_server`
- `TestDenyPaths::test_revoked_token_fails`
- `TestCrossTransportConsistency::test_get_system_time_both_transports`
- `TestCrossTransportConsistency::test_convert_time_both_transports`

#### `test-mcp-protocol-e2e` — 19 erroring

All 19 error in the `client` fixture setup:
- `TestConnectivity::test_ping`
- `TestConnectivity::test_initialize_reports_server_info`
- `TestConnectivity::test_server_capabilities_include_core_surfaces`
- `TestConnectivity::test_multiple_calls_in_one_session`
- `TestTools::test_tools_list_nonempty`
- `TestTools::test_tools_have_required_fields`
- `TestTools::test_tools_include_gateway_prefixed`
- `TestTools::test_tool_input_schemas_are_json_schema_objects`
- `TestDiscovery::test_resources_list`
- `TestDiscovery::test_resources_read_roundtrip`
- `TestDiscovery::test_prompts_list`
- `TestDiscovery::test_prompt_get_renders`
- `TestToolCalls::test_get_system_time`
- `TestToolCalls::test_echo`
- `TestToolCalls::test_convert_time`
- `TestToolCalls::test_get_stats`
- `TestToolCalls::test_schema_error_preserves_payload`
- `TestToolCalls::test_schema_success_validates_payload`
- `TestToolCalls::test_nonexistent_tool`

Skipped (1): `test_mcp_protocol_e2e.py:411` — Rust MCP public transport not active (expected; Rust mode off).
Passed (2): unauthenticated `/mcp/` → 401, and invalid method → 200 (the only checks that don't open an MCP session).

</details>

### Affected tests
- `tests/live_gateway/mcp/test_mcp_rbac_transport.py`
- `tests/live_gateway/mcp/test_mcp_protocol_e2e.py`

### PR DESCRIPTION
**Pull request by Yosiefeyob**
## 🔗 Related Issue

Closes #5162

---

## 📝 Summary

The `test-mcp-rbac` and `test-mcp-protocol-e2e` gates failed deterministically on the testing stack. Every test that opened an MCP session died at the streamable HTTP `/mcp/` initialize handshake with:

```text
Failed to initialize server session
```

This surfaced from FastMCP at `client.py:532`.

---

## 🧩 Root Cause

The test client connects via `localhost`, which can resolve to IPv6 (`::1`) first. However, the docker-compose testing stack publishes the gateway port on IPv4 only.

As a result, the async FastMCP/httpx client stalls on the unreachable `::1` address until the initialization timeout fires. The sync `httpx.get` reachability probe does not hit this path, so the suite reports errors instead of skipping.

This reproduces on 1, 2, and 3 replicas, so it is not replica-count or affinity specific.

---

## 🔧 Change

- Updated `tests/live_gateway/helpers/mcp_test_helpers.py` to normalize a `localhost` host in `BASE_URL` to `127.0.0.1` using `_prefer_ipv4_localhost`.
- This works whether the URL comes from the default value or from `MCP_CLI_BASE_URL`.
- This matters because the documented reproduction steps set `MCP_CLI_BASE_URL` to `http://localhost:8080`.
- Updated the `Makefile` so the `test-mcp-*` status messages match the IPv4 default.

This is test-only. It does not change product code, gateway behavior, nginx, Docker, or infrastructure.

---

## 🏷️ Type of Change

- [x] Bug fix
- [ ] Feature / Enhancement
- [ ] Documentation
- [ ] Refactor
- [ ] Chore (deps, CI, tooling)
- [ ] Other (describe below)

---

## 🧪 Verification

- `test-mcp-protocol-e2e`: 19 passed / 3 skipped  
  Previously: 19 errors
- `test-mcp-rbac`: 40 passed

| Check | Command | Status |
|---|---|---|
| MCP protocol e2e | `make test-mcp-protocol-e2e` | ✅ Passed |
| MCP RBAC | `make test-mcp-rbac` | ✅ Passed |

---

## ✅ Checklist

- [ ] Code formatted (`make black isort pre-commit`)
- [x] Tests added/updated for changes
- [ ] Documentation updated (if applicable)
- [x] No secrets or credentials committed

---

## 📓 Notes

This PR pins the live-gateway MCP test client to IPv4 when the configured URL uses `localhost`, avoiding IPv6-first resolution to `::1` in environments where the docker-compose gateway port is only reachable via IPv4.

### CODE REVIEW

**Review by gcgoncalves — DISMISSED** [review]

File: tests/live_gateway/helpers/mcp_test_helpers.py [hunk 1]
```diff
@@ -29,14 +29,45 @@
...
+
 # ---------------------------------------------------------------------------
 # Configuration
 # ---------------------------------------------------------------------------
-BASE_URL = os.getenv("MCP_CLI_BASE_URL", "http://127.0.0.1:8080")  # IPv4 explicit — `localhost` resolves IPv6-first under the test conftest's getaddrinfo stub, which docker-compose doesn't bind
+def _prefer_ipv4_localhost(url: str) -> str:
+    """Rewrite a ``localhost`` host in ``url`` to ``127.0.0.1``.
+
+    ``localhost`` resolves IPv6-first (``::1``) under the test conftest's
+    ``getaddrinfo`` stub (and on many real hosts), but docker-compose only
```
Comment (cafalchio) [inline tests/live_gateway/helpers/mcp_test_helpers.py]: Please simplify the doc.

**Comment by ja8zyjits** [comment]
## Summary

Test-only fix. Normalizes `localhost` → `127.0.0.1` in MCP test helpers to avoid IPv6 resolution stalls. Fixes 46 failing tests (27 `test-mcp-rbac`, 19 `test-mcp-protocol-e2e`).

**Changed Files**:
- `.secrets.baseline` - timestamp only
- `Makefile` - echo messages updated to `127.0.0.1:8080`
- `tests/live_gateway/helpers/mcp_test_helpers.py` - core fix

**Root Cause**: Docker-compose binds IPv4 only. Async MCP client tries IPv6 first for `localhost`, stalls. Fix pins to `127.0.0.1`.

---

## 🚨 Blocking Issues

### Missing Test Coverage

**Priority 1 (Blocking)**

- [ ] **No unit tests for `_prefer_ipv4_localhost()` helper**
  - File: `tests/live_gateway/helpers/mcp_test_helpers.py:42-68`
  - Risk: Critical path function, zero coverage
  - Create: `tests/unit/live_gateway/helpers/test_mcp_test_helpers.py`
  - Required tests (8):
    - `test_prefer_ipv4_localhost_rewrites_localhost()` - `http://localhost:8080` → `http://127.0.0.1:8080`
    - `test_prefer_ipv4_localhost_preserves_ipv4()` - `http://127.0.0.1:8080` unchanged
    - `test_prefer_ipv4_localhost_preserves_ipv6()` - `http://[::1]:8080` unchanged
    - `test_prefer_ipv4_localhost_preserves_domain()` - `http://example.com:8080` unchanged
    - `test_prefer_ipv4_localhost_handles_userinfo()` - `http://user:pass@localhost:8080` → `http://user:pass@127.0.0.1:8080`
    - `test_prefer_ipv4_localhost_handles_no_port()` - `http://localhost` → `http://127.0.0.1`
    - `test_prefer_ipv4_localhost_handles_https()` - `https://localhost:443` → `https://127.0.0.1:443`
    - `test_prefer_ipv4_localhost_handles_path_query_fragment()` - preserves path/query/fragment

- [ ] **Edge case: malformed URLs not tested**
  - File: `tests/live_gateway/helpers/mcp_test_helpers.py:42`
  - Risk: `urlsplit()` succeeds on malformed input, unexpected results
  - Add tests:
    - Empty string: `_prefer_ipv4_localhost("")`
    - Invalid scheme: `_prefer_ipv4_localhost("ftp://localhost:8080")`
    - No scheme: `_prefer_ipv4_localhost("localhost:8080")`
    - IPv6 zone ID: `_prefer_ipv4_localhost("http://[fe80::1%eth0]:8080")`

**Priority 2 (High)**

- [ ] **No integration test verifying fix resolves #5162**
  - Current: No regression test that would fail without fix
  - Add to: `tests/integration/live_gateway/test_mcp_ipv4_localhost_fix.py` or `test_mcp_protocol_e2e.py`
  - Test should:
    1. Mock `getaddrinfo` to return IPv6 first for `localhost`
    2. Verify `BASE_URL` uses `127.0.0.1` after normalization
    3. Confirm MCP client initializes successfully

- [ ] **No E2E verification of normalization**
  - Files affected: All `tests/live_gateway/mcp/`
  - Add smoke test to `test_mcp_protocol_e2e.py::TestConnectivity`:
    - `MCP_CLI_BASE_URL=http://localhost:8080` → `BASE_URL` becomes `http://127.0.0.1:8080`
    - MCP client initializes (no timeout)

- [ ] **No test for `MCP_CLI_BASE_URL` env var override**
  - File: `tests/live_gateway/helpers/mcp_test_helpers.py:70`
  - Test: Set `MCP_CLI_BASE_URL=http://localhost:9999`, verify result is `http://127.0.0.1:9999`

**Priority 3 (Medium)**

- [ ] **Add error handling for malformed URLs**
  - Consider: What if `MCP_CLI_BASE_URL` set to invalid value?

---

## ⚠️ Warnings

- [ ] **Makefile echo vs docs inconsistency**
  - Makefile: `127.0.0.1:8080` (lines 814,850,858,863,868,873,878,891)
  - Docs: `localhost:8080` (44 files, 155 occurrences in `docs/docs/`)
  - Impact: Minor cosmetic inconsistency
  - Action: Optional follow-up to align docs or keep `localhost` (both work due to normalization)

- [ ] **Test helper not documented**
  - File: `tests/live_gateway/helpers/mcp_test_helpers.py` (`_prefer_ipv4_localhost()`)
  - Missing from: `docs/docs/testing/index.md`, `docs/docs/testing/LocalTest.md`
  - Add note explaining:
    - `MCP_CLI_BASE_URL` defaults to `http://127.0.0.1:8080`
    - `localhost` auto-normalized to `127.0.0.1` (avoids IPv6 issues)
    - Matters for docker-compose IPv4-only stacks

---

## ✅ Good Practices

- [x] **Test-only change** - PR description clear, verified
- [x] **Root cause documented** - `issue_5162_pr_5230.md` has excellent technical detail
- [x] **Idempotent, safe** - Only rewrites `localhost`, preserves other URLs
- [x] **Comprehensive docstring** - Explains problem, root cause, solution, references #5162
- [x] **Minimal, focused** - Test infrastructure only, no product code
- [x] **Makefile consistency** - Echo messages match actual behavior
- [x] **Proper URL parsing** - Uses stdlib `urlsplit`/`urlunsplit`, handles edge cases

---

## Test Standards Compliance

| Standard | Status | Notes |
|----------|--------|-------|
| No synthetic tests | ✅ Pass | No new tests added |
| Edge cases covered | ❌ Fail | Helper lacks edge case tests |
| Security deny-path | N/A | Test infrastructure change |
| Integration tests | ❌ Fail | No test verifies fix resolves #5162 |
| Tests verify behavior | N/A | No new tests to evaluate |
| Tests independent | N/A | No new tests to evaluate |

---

## Code Quality

**No issues.** Clean, well-documented, follows best practices:

1. **Logic**: Correct URL rewriting, proper edge case handling
2. **Type Safety**: Complete type hints (`str -> str`)
3. **Documentation**: Excellent docstring
4. **Maintainability**: Simple, single responsibility
5. **Error Handling**: Appropriate for connectivity checks

Makefile changes mechanical (cosmetic echo updates).

---

## Recommendation

**REQUEST_CHANGES** - Add unit tests for `_prefer_ipv4_localhost()` before merge. Consider integration test to verify fix and prevent regressions.

---

<details>
<summary>Agent Details: Code Docs</summary>

**Documentation Accuracy**: 95% aligned

No blocking issues. Main gap: Makefile shows `127.0.0.1:8080`, docs show `localhost:8080`. Minor cosmetic inconsistency. Test helper normalization not documented in testing guides.

**Optional follow-up**:
1. Add note in `docs/docs/testing/index.md` explaining IPv4 normalization
2. Decide: update docs to `127.0.0.1:8080` or keep `localhost:8080` (both work)

</details>

<details>
<summary>Agent Details: Test Quality</summary>

**Summary**: Fix sound, well-documented, **lacks test coverage**.

`_prefer_ipv4_localhost()` is critical infrastructure affecting all MCP live-gateway tests, yet has zero unit tests. Risk for future refactoring/edge case bugs.

**Recommendation**: Add unit tests before merge. Add integration test to verify fix resolves #5162.

</details>

<details>
<summary>Agent Details: Code Quality</summary>

**No issues found.**

- Logic: ✅ Correct
- Type Safety: ✅ Complete hints
- Documentation: ✅ Excellent docstring
- Maintainability: ✅ Simple, focused
- Error Handling: ✅ Appropriate

Fix addresses IPv6 resolution issue in minimal, targeted way without complexity or technical debt.

</details>

**Review by cafalchio — DISMISSED** [review]
Great one. Simple and it does the job.

**Review by ja8zyjits — DISMISSED** [review]
LGTM

**Review by ja8zyjits — DISMISSED** [review]
LGTM
process followed

**Review by ja8zyjits — APPROVED** [review]
LGTM

**Review by cafalchio — APPROVED** [review]
LGTM
